src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java

```diff
@@ -0,0 +1,420 @@
+            INVALID_BOUNDARY_FORMAT);
+
+    private static boolean isRelationTypeBoundaryWithBoundaryTag(final AtlasObject object)
+    {
+        return Validators.isOfType(object, RelationTypeTag.class, RelationTypeTag.BOUNDARY)
+                && Validators.hasValuesFor(object, BoundaryTag.class);
+    }
+
+    public BoundaryIntersectionCheck(final Configuration configuration)
+    {
+        super(configuration);
+    }
+
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return object instanceof Relation && isRelationTypeBoundaryWithBoundaryTag(object);
+    }
+
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Map<String, Relation> tagToRelation = this.getRelationMap(object);
+        final RelationBoundary relationBoundary = new RelationBoundary(tagToRelation,
+                this.getBoundaryAtlasEntities((Relation) object));
+        final Set<String> instructions = new HashSet<>();
+        final Set<AtlasObject> objectsToFlag = new HashSet<>();
+        final Set<String> matchedTags = new HashSet<>();
+        final Atlas atlas = object.getAtlas();
+        for (final AtlasEntity atlasEntity : relationBoundary.getAtlasEntities())
+        {
+            final Iterable<LineItem> lineItemsIntersecting = atlas.lineItemsIntersecting(
+                    atlasEntity.bounds(), this.getPredicateForLineItemSelection(atlasEntity,
+                            relationBoundary.getTagToRelation().keySet()));
+            final Iterable<Area> areasIntersecting = atlas.areasIntersecting(atlasEntity.bounds(),
+                    this.getPredicateForAreaSelection(atlasEntity,
+                            relationBoundary.getTagToRelation().keySet()));
+            final Set<String> currentMatchedTags = new HashSet<>();
+            matchedTags.addAll(this.processLineItems(relationBoundary, instructions, objectsToFlag,
+                    atlasEntity, lineItemsIntersecting, currentMatchedTags));
+            matchedTags.addAll(this.processAreas(relationBoundary, instructions, objectsToFlag,
+                    atlasEntity, areasIntersecting, currentMatchedTags));
+
+            objectsToFlag.addAll(relationBoundary.getRelationsByBoundaryTags(matchedTags));
+            objectsToFlag.add(atlasEntity);
+        }
+        if (instructions.isEmpty())
+        {
+            return Optional.empty();
+        }
+        else
+        {
+            final CheckFlag checkFlag = new CheckFlag(this.getTaskIdentifier(object));
+            instructions.forEach(checkFlag::addInstruction);
+            checkFlag.addObjects(objectsToFlag);
+            return Optional.of(checkFlag);
+        }
+    }
+
+    @Override
+    protected List<String> getFallbackInstructions()
+    {
+        return FALLBACK_INSTRUCTIONS;
+    }
+
+    private boolean checkAtlasEntityAsBoundary(final AtlasEntity atlasEntity,
```

Comment:
It would be great if you could add java doc to each of the methods added in the class with a description of what each method is doing.